actgen/agents/dqn.py

```diff
@@ -79,16 +81,33 @@ def _get_epsilon(self, testing=False):
     def _get_q_targets(self, batch):
         with torch.no_grad():
             # Compute Double-Q targets
-            next_state = torch.stack(batch.next_state).float()
-            ap = torch.argmax(self.q(next_state), dim=-1)
-            vp = self.q_target(next_state).gather(-1, ap.unsqueeze(-1)).squeeze(-1)
-            not_done_idx = ~torch.stack(batch.done)
-            targets = torch.stack(batch.reward) + self.params['gamma'] * vp * not_done_idx
+            next_state = torch.stack(batch.next_state).float()  # (batch_size, dim_state)
+            ap = torch.argmax(self.q(next_state), dim=-1)  # (batch_size, )
+            vp = self.q_target(next_state).gather(-1, ap.unsqueeze(-1)).squeeze(-1)  # (batch_size, )
+            not_done_idx = ~torch.stack(batch.done)  # (batch_size, )
+            targets = torch.stack(batch.reward) + self.params['gamma'] * vp * not_done_idx  # (batch_size, )
+            if self.params['dqn_train_pin_other_q_values']:
+                all_action_targets = self._get_q_predictions(batch)  # (batch_size, n_actions)
+                # for sample in batch
+                for i, target in enumerate(targets):
+                    action_taken = batch.action[i]
+                    if self.params['oracle']:
+                        # oracle update
+                        num_dup = self.params['duplicate']
+                        original_a = math.floor(action_taken / num_dup)
+                        similar_a = list(range(original_a * num_dup, original_a * num_dup + num_dup))
+                        all_action_targets[i, similar_a] = target
+                    else:
+                        # normal update
+                        all_action_targets[i, action_taken] = target
+                return all_action_targets
         return targets
 
     def _get_q_predictions(self, batch):
-        q_values = self.q(torch.stack(batch.state).float())
-        q_acted = extract(q_values, idx=torch.stack(batch.action).long(), idx_dim=-1)
+        q_values = self.q(torch.stack(batch.state).float())  # (batch_size, n_actions)
```

Comment:
I keep seeing you use `float()` and `long()` everywhere. Is there a reason you're doing that?

I wonder if there's a way we can avoid casting in so many different places...

Reply:
It's because I saw you always used `float()` in forward-pass calls in the scaffolding of the DQN code. I'm not sure if we need to cast it, because I assume the default dtype in pytorch would be floats.

Reply:
No, that's right. In general whenever you convert from numpy to pytorch, you need to be specific about what type you want the tensors to have. The numpy default is float64, but the pytorch default is float32, and you'll get type errors if you mix and match.

But my question is about whether there's a way to avoid converting so often.

Reply:
maybe we could do `torch.set_default_dtype(torch.float64)` at the very beginning of the script?

Reply:
I believe that will only set the default for newly constructed tensors, not tensors that are converted from numpy.